File: app/src/main/java/com/majeur/psclient/model/Move.java
package com.majeur.psclient.model;

import org.json.JSONException;
import org.json.JSONObject;

import androidx.annotation.Nullable;

import static com.majeur.psclient.model.Id.toId;

public class Move {

    public final String name;
    public final String id;
    public final int index;
    public final int pp;
    public final int ppMax;
    public final Target target;
    public final boolean disabled;
    public final String zName;
    public Details details;
    public Details zDetails;

    public String maxMoveId;
    public Target maxMoveTarget;
    public Details maxDetails;

    // Flag to know if this move should be read as
    // zmove or regular base one
    public boolean zflag;
    public boolean maxflag;

    public Move(int index, JSONObject jsonObject, @Nullable JSONObject zJsonObject, JSONObject maxJsonObject) throws JSONException {
        this.index = index;
        name = jsonObject.getString("move").replace("Hidden Power", "HP");
        id = jsonObject.getString("id");
        pp = jsonObject.optInt("pp", -1);
        ppMax = jsonObject.optInt("maxpp", -1);
        target = Target.parse(jsonObject.optString("target", null));
        disabled = jsonObject.optBoolean("disabled", false);
        zName = zJsonObject != null ? zJsonObject.optString("move", null) : null;
        maxMoveId = maxJsonObject != null ? maxJsonObject.optString("move", null) : null;
        maxMoveTarget = maxJsonObject != null ? Target.parse(maxJsonObject.optString("target", null)) : null;
    }

    public boolean canZMove() {
        return zName != null;
    }

    public String maxMoveName() {
        for (String name : MAX_MOVES)
            if (toId(name).equals(maxMoveId))
                return name;
        return maxMoveId;
//        switch (details.type) {
//                case "Poison":  return "Max Ooze";
//                case "Fighting": return "Max Knuckle";
//                case "Dark": return "Max Darkness";
//                case "Grass": return "Max Overgrowth";
//                case "Normal": return "Max Strike";
//                case "Rock": return "Max Rockfall";
//                case "Steel": return "Max Steelspike";
//                case "Dragon": return "Max Wyrmwind";
//                case "Electric": return "Max Lightning";
//                case "Water": return "Max Geyser";
//                case "Fire": return "Max Flare";
//                case "Ghost": return "Max Phantasm";
//                case "Bug": return "Max Flutterby";
//                case "Psychic": return "Max Mindstorm";
//                case "Ice": return "Max Hailstorm";
//                case "Flying": return "Max Airstream";
//                case "Ground": return "Max Quake";
//                case "Fairy": return "Max Starfall";
//                case "???": return "";
//                default: return null;
//        }
    }

    private static String[] MAX_MOVES = {"Max Guard", "Max Ooze", "Max Knuckle", "Max Darkness",
            "Max Overgrowth", "Max Strike", "Max Rockfall", "Max Steelspike", "Max Wyrmwind",
            "Max Lightning", "Max Geyser", "Max Flare", "Max Phantasm", "Max Flutterby", "Max Mindstorm",
            "Max Hailstorm", "Max Airstream", "Max Quake", "Max Starfall"};

    @Override
    public String toString() {
        return "Move{" +
                "name='" + name + '\'' +
                ", id='" + id + '\'' +
                ", index=" + index +
                ", pp=" + pp +
                ", ppMax=" + ppMax +
                ", target='" + target + '\'' +
                ", disabled=" + disabled +
                ", zName='" + zName + '\'' +
                ", details=" + details +
                '}';
    }

    public static class Details {

        public final String name;
        public final int accuracy;
        public final int priority;
        public final int basePower;
        public final int zPower;
        public final int color;
        public final String category;
        public final String desc;
        public final String type;
        public final int pp;
        public final Target target;
        public final String zEffect;
        public final int maxPower;

        public Details(String name, int accuracy, int priority, int basePower, int zPower,
                       String category, String desc, String type, int pp, String target,
                       String zEffect, int maxPower) {
            this.name = name;
            this.accuracy = accuracy;
            this.priority = priority;
            this.basePower = basePower;
            this.zPower = zPower;
            this.category = category;
            this.desc = desc;
            this.type = type;
            this.color = Colors.typeColor(toId(type));
            this.pp = pp;
            this.target = Target.parse(target);
            this.zEffect = zEffect;
            this.maxPower = maxPower;
        }

        @Override
        public String toString() {
            return "Details{" +
                    "name='" + name + '\'' +
                    ", accuracy=" + accuracy +
                    ", priority=" + priority +
                    ", basePower=" + basePower +
                    ", zPower=" + zPower +
                    ", color=" + color +
                    ", category='" + category + '\'' +
                    ", desc='" + desc + '\'' +
                    ", type='" + type + '\'' +
                    ", pp=" + pp +
                    ", target=" + target +
                    ", zEffect='" + zEffect + '\'' +
                    '}';
        }
    }

    public enum Target {

        NORMAL,
        ALL_ADJACENT_FOES,
        SELF,
        ANY,
        ADJACENT_ALLY_OR_SELF,
        ALLY_TEAM,
        ADJACENT_ALLY,
        ALLY_SIDE,
        ALL_ADJACENT,
        SCRIPTED,
        ALL,
        ADJACENT_FOE,
        RANDOM_NORMAL,
        FOE_SIDE;

        public static Target parse(String target) {
            if (target == null) return Target.NORMAL;
            target = target.toLowerCase().trim();
            switch (target) {
                case "normal": return Target.NORMAL;
                case "alladjacentfoes": return Target.ALL_ADJACENT_FOES;
                case "self": return Target.SELF;
                case "any": return Target.ANY;
                case "adjacentallyorself": return Target.ADJACENT_ALLY_OR_SELF;
                case "allyteam": return Target.ALLY_TEAM;
                case "adjacentally": return Target.ADJACENT_ALLY;
                case "allyside": return Target.ALLY_SIDE;
                case "alladjacent": return Target.ALL_ADJACENT;
                case "scripted": return Target.SCRIPTED;
                case "all": return Target.ALL;
                case "adjacentfoe": return Target.ADJACENT_FOE;
                case "randomnormal": return Target.RANDOM_NORMAL;
                case "foeside": return Target.FOE_SIDE;
                default: return Target.NORMAL;
            }
        }


        //  Triples       Doubles     Singles
        //  3  2  1         2  1         1
        // -1 -2 -3        -1 -2        -1
        public static boolean[][] computeTargetAvailabilities(Target target, int position, int pokeCount) {
            boolean[][] availabilities = new boolean[2][pokeCount];
            for (int i = 0; i < 2; i++)
                for (int j = 0; j < pokeCount; j++) availabilities[i][j] = true;

            if (target == ADJACENT_FOE) {
                for (int i = 0; i < pokeCount; i++) {
                    if (position != i && position != i + 1 && position != i - 1) availabilities[0][i] = false;
                    availabilities[1][i] = false;
                }
            }
            if (target == ADJACENT_ALLY || target == ADJACENT_ALLY_OR_SELF) {
                for (int i = 0; i < pokeCount; i++) {
                    if (position != i && position != i + 1 && position != i - 1) availabilities[1][i] = false;
                    availabilities[0][i] = false;
                }
                if (target == ADJACENT_ALLY) availabilities[1][position] = false;
            }
            if (target == NORMAL) {
                availabilities[1][position] = false;
            }
            return availabilities;
        }

        public boolean isChosable() {
            return this == NORMAL || this == ANY || this == ADJACENT_ALLY
                    || this == ADJACENT_ALLY_OR_SELF || this == ADJACENT_FOE;
        }
    }
}

File: app/src/main/java/com/majeur/psclient/model/battle/Move.kt
package com.majeur.psclient.model.battle

import com.majeur.psclient.model.common.Colors.typeColor
import com.majeur.psclient.util.toId
import org.json.JSONObject

class Move(val index: Int, json: JSONObject, zJson: JSONObject?, maxJson: JSONObject?) {

    val name = json.getString("move").replace("Hidden Power", "HP")
    val id: String = json.getString("id")
    val pp = json.optInt("pp", -1)
    val ppMax = json.optInt("maxpp", -1)
    val target = Target.parse(json.optString("target"))
    val disabled = json.optBoolean("disabled")
    val zName = zJson?.optString("move")
    var details: Details? = null
    var zDetails: Details? = null
    var maxMoveId = maxJson?.optString("move")
    var maxMoveTarget = if (maxJson != null) Target.parse(maxJson.optString("target")) else null
    var maxDetails: Details? = null

    // Flag to know if this move should be read as
    // z-move, max-move or regular move
    var zflag = false
    var maxflag = false

    val canZMove get() = zName != null

    val maxMoveName: String?
        get() = MAX_MOVES.firstOrNull { it.toId() == maxMoveId } ?: maxMoveId

    class Details {

        lateinit var name: String
        lateinit var category: String
        var desc: String? = null
        var accuracy = 0
        var priority = 0
        var basePower = 0
        var zPower = 0
        var color = 0
        var pp = 0
        var type: String? = null
            set(value) {
                field = value
                color = typeColor(value?.toId())
            }
        var target: Target? = null
        var zEffect: String? = null
            set(value) { field = zMoveEffects(value)
            }

        var maxPower = 0

        companion object {

            private fun zMoveEffects(effect: String?) = when (effect) {
                "clearnegativeboost" -> "Restores negative stat stages to 0"
                "crit2" -> "Crit ratio +2"
                "heal" -> "Restores HP 100%"
                "curse" -> "Restores HP 100% if user is Ghost type, otherwise Attack +1"
                "redirect" -> "Redirects opposing attacks to user"
                "healreplacement" -> "Restores replacement's HP 100%"
                else -> null

            }

        }
    }

    enum class Target {

        NORMAL, ALL_ADJACENT_FOES, SELF, ANY, ADJACENT_ALLY_OR_SELF, ALLY_TEAM, ADJACENT_ALLY,
        ALLY_SIDE, ALL_ADJACENT, SCRIPTED, ALL, ADJACENT_FOE, RANDOM_NORMAL, FOE_SIDE;

        val isChoosable: Boolean
            get() = this == NORMAL || this == ANY || this == ADJACENT_ALLY || this == ADJACENT_ALLY_OR_SELF
                    || this == ADJACENT_FOE

        companion object {

            fun parse(target: String) = when (target.toId()) {
                "normal" -> NORMAL
                "alladjacentfoes" -> ALL_ADJACENT_FOES
                "self" -> SELF
                "any" -> ANY
                "adjacentallyorself" -> ADJACENT_ALLY_OR_SELF
                "allyteam" -> ALLY_TEAM
                "adjacentally" -> ADJACENT_ALLY
                "allyside" -> ALLY_SIDE
                "alladjacent" -> ALL_ADJACENT
                "scripted" -> SCRIPTED
                "all" -> ALL
                "adjacentfoe" -> ADJACENT_FOE
                "randomnormal" -> RANDOM_NORMAL
                "foeside" -> FOE_SIDE
                else -> NORMAL
            }

            //  Triples       Doubles     Singles
            //  3  2  1         2  1         1
            // -1 -2 -3        -1 -2        -1
            fun computeTargetAvailabilities(target: Target, position: Int, pokeCount: Int): Array<BooleanArray> {
                val availabilities = Array(2) { BooleanArray(pokeCount) }
                for (i in 0..1) for (j in 0 until pokeCount) availabilities[i][j] = true
                if (target == ADJACENT_FOE) {
                    for (i in 0 until pokeCount) {
                        if (position != i && position != i + 1 && position != i - 1) availabilities[0][i] = false
                        availabilities[1][i] = false
                    }
                }
                if (target == ADJACENT_ALLY || target == ADJACENT_ALLY_OR_SELF) {
                    for (i in 0 until pokeCount) {
                        if (position != i && position != i + 1 && position != i - 1) availabilities[1][i] = false
                        availabilities[0][i] = false
                    }
                    if (target == ADJACENT_ALLY) availabilities[1][position] = false
                }
                if (target == NORMAL) {
                    availabilities[1][position] = false
                }
                return availabilities
            }
        }
    }

    companion object {
        private val MAX_MOVES = arrayOf("Max Guard", "Max Ooze", "Max Knuckle", "Max Darkness",
                "Max Overgrowth", "Max Strike", "Max Rockfall", "Max Steelspike", "Max Wyrmwind",
                "Max Lightning", "Max Geyser", "Max Flare", "Max Phantasm", "Max Flutterby", "Max Mindstorm",
                "Max Hailstorm", "Max Airstream", "Max Quake", "Max Starfall")
    }
}